dask_expr/_merge.py

```diff
@@ -428,9 +428,20 @@ def _layer(self) -> dict:
         from distributed.shuffle._shuffle import shuffle_barrier
 
         dsk = {}
-        token = self._name.split("-")[-1]
-        token_left = token + "-left"
-        token_right = token + "-right"
+        token_left = _tokenize_deterministic(
+            "hash-join",
```

Comment:
Added this to ensure that we avoid any unwanted sharing between merges and shuffles. I don't have a good-enough overview right now to guarantee that sharing between them works.

Reply:
If we can rely on `name_input_{left/right}` to truly be unique and identify distinct input dataframes, I believe we can even share between merges and ordinary shuffles. I wonder where this use case would come up, though

Reply:
My main concern here is that `shuffle_transfer` and `merge_transfer` do different things. We'd have to refactor this so that they are in fact interchangeable and shareable. I guess there are only few cases where this would come in handy though. Maybe something like joining a dataframe on `x` and aggregating that same dataframe on `x` within the same graph?